File: medium/javascript/1679-max-number-of-k-sum-pairs.js
/**
 * O(n)
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function (nums, k) {
  const map = new Map();
  let operations = 0;

  for (let i = 0; i < nums.length; i++) {
    const dif = k - nums[i];
    const mappedDif = map.get(dif);

    if (mappedDif !== undefined && mappedDif.length > 0) {
      operations++;
      mappedDif.pop();
    } else {
      const mapped = map.get(nums[i]);
      mapped ? mapped.push(i) : map.set(nums[i], [i]);
    }
  }

  return operations;
};

File: medium/go/1679-max-number-of-k-sum-pairs.go
package main

// O(n)
func maxOperations(nums []int, k int) int {
	numMap := make(map[int]int, len(nums))
	operations := 0

	for _, v := range nums {
		if v >= k {
			continue
		}

		dif := k - v

		if numMap[dif] != 0 {
			numMap[dif]--
			operations++
			continue
		}
		numMap[v]++
    }
 
	return operations 
}